port_ocean/clients/port/mixins/entities.py

```diff
@@ -169,17 +316,103 @@ async def batch_upsert_entities(
 
+    async def upsert_entities_in_batches(
+        self,
+        entities: list[Entity],
+        request_options: RequestOptions,
+        user_agent_type: UserAgentType | None = None,
+        should_raise: bool = True,
+    ) -> list[tuple[bool, Entity]]:
+        """
+        This function upserts a list of entities into Port in batches.
+        The batch size is calculated based on both the number of entities and their size.
+        Batches are processed in parallel using asyncio.gather, with concurrency controlled by the semaphore.
+
+        :param entities: A list of Entities to be upserted
+        :param request_options: A dictionary specifying how to upsert the entity
+        :param user_agent_type: a UserAgentType specifying who is preforming the action
+        :param should_raise: A boolean specifying whether the error should be raised or handled silently
+        :return: A list of tuples where each tuple contains:
+            - First value: True if entity was created successfully, False if there was an error
+            - Second value: The reduced entity with updated identifier (if successful) or the original entity (if failed)
+        """
+        entities_results: list[tuple[bool, Entity]] = []
+        blueprint = entities[0].blueprint
+
+        if ocean.config.bulk_upserts_enabled:
+            batch_size = self.calculate_entities_batch_size(entities)
+            batches = [
+                entities[i : i + batch_size]
+                for i in range(0, len(entities), batch_size)
+            ]
+
+            batch_results = await asyncio.gather(
+                *(
+                    self.upsert_entities_batch(
+                        blueprint,
+                        batch,
+                        request_options,
+                        user_agent_type,
+                        should_raise=should_raise,
+                    )
+                    for batch in batches
+                ),
+                return_exceptions=True,
+            )
+
+            for batch, batch_result in zip(batches, batch_results):
+                if isinstance(batch_result, httpx.HTTPStatusError) or isinstance(
+                    batch_result, Exception
+                ):
+                    if should_raise:
+                        raise batch_result
+                    # If should_raise is False, retry batch in sequential order as a fallback only for 413 errors
+                    if (
+                        isinstance(batch_result, httpx.HTTPStatusError)
+                        and batch_result.response.status_code == 413
+                    ):
+                        sequential_results = (
+                            await self._upsert_entities_batch_sequential(
+                                batch, request_options, user_agent_type, should_raise
+                            )
+                        )
+                        entities_results.extend(sequential_results)
+                    else:
```

Comment:
I wonder if we should accumulate all the failed ones on that error and then respawn all in the sequential.

Reply:
similar to line 409